sections/docker/avoid-build-time-secrets.md

```diff
@@ -0,0 +1,92 @@
+# Clean build-time secrets, avoid secrets as args
+
+<br/><br/>
+
+### One Paragraph Explainer
+
+A docker image is not just a bunch of files rather constitutes multiple layers that reveal also what happened during the build time. In a very common scenario, developers need the npm token during build time (mostly for private registries) - this is falsely achieved by passing the token as a build time args. It might seem innocent and safe, however in fact this token can now be fetched from the developers machine Docker history, from the Docker registry and the CI. An attacker who get access to that token, is now capable of writing into the organization private npm registry. There are two more secured alternatives: The flawless one is using Docker --secret feature (experimental as of July 2020) which allows mounting a file during build time only. The second approach is using multi-stage build with args, building and then copying only the necessary files to production. The last technique will not prevent the secrets file from appearing in the local docker history but considered as as secured enough in some organizations.
+
+<br/><br/>
+
+### Code Example – Using Docker mounted secrets (experimental but stable)
+
+<details>
+
+<summary><strong>Dockerfile</strong></summary>
+
+```
+# syntax = docker/dockerfile:1.0-experimental
+
+FROM node:12-slim
+WORKDIR /usr/src/app
+COPY package.json package-lock.json ./
+RUN --mount=type=secret,id=npm,target=/root/.npmrc npm ci
+
+# The rest comes here
+```
+
+</details>
+
+<br/><br/>
+
+### Code Example – Building securely using multi-stage build
+
+<details>
+
+<summary><strong>Dockerfile</strong></summary>
+
+```
+
+FROM node:12-slim AS build
+ARG NPM_TOKEN
+WORKDIR /usr/src/app
+COPY . /dist
+RUN echo "//registry.npmjs.org/:\_authToken=\$NPM_TOKEN" > .npmrc && \
+ npm ci --production && \
+ rm -f .npmrc
+
+FROM build as prod
+COPY --from=build /dist /dist
+CMD ["node","index.js"]
+
+# The ARG and .npmrc won't appear in the final image, but can be found in the Docker daemon un-tagged images list - make sure to delete those
+```
+
+</details>
+
+<br/><br/>
+
+### Code Example Anti Pattern – Using build time args
+
+<details>
+
+<summary><strong>Dockerfile</strong></summary>
+
+```
+
+FROM node:12-slim
+ARG NPM_TOKEN
+WORKDIR /usr/src/app
+COPY . /dist
+RUN echo "//registry.npmjs.org/:\_authToken=\$NPM_TOKEN" > .npmrc && \
+ npm ci --production && \
+ rm -f .npmrc
+
+# Deleting the .npmrc within the same copy command will not save it inside the layer, however it can be found in image history
+
+CMD ["node","index.js"]
+```
+
+</details>
+
+<br/><br/>
+
+### Blog Quote: "These secrets aren’t saved in the final Docker"

```

Comment:
It sounds like there's something missing here 😄 Not completely sure but I think this talks about the image, so it might be "These secrets aren't saved in the final Docker image", but maybe it's also about the container? 🤔